app/controllers/admin/categories_controller.rb

```diff
@@ -0,0 +1,72 @@
+class Admin::CategoriesController < ApplicationController
+  before_action :load_category, only: %i(edit update destroy)
+  before_action :load_parent_category, only: %i(new edit)
+
+  def index
+    @categories = Category.parent_cats.paginate page: params[:page]
+  end
+
+  def new
+    @category = Category.new
+  end
+
+  def create
+    @category = Category.new category_params
+    @category.parent_id = 0 if @category.parent_id.nil?
+    if @category.save
+      flash[:success] = t ".success"
+      redirect_to admin_categories_path
+    else
+      flash[:danger] = t ".fail"
+      render :new
+    end
+  end
+
+  def show; end
+
+  def edit; end
+
+  def update; end
+
+  def destroy
+    if @category.parent_id.zero?
+      delete_parent_category @category
+    elsif @category.destroy
+      flash[:success] = t ".success"
+    else
+      flash[:danger] = t ".fail"
+    end
+    redirect_to request.referer
+  end
+
+  private
+  def category_params
+    params.require(:category).permit(:name, :parent_id)
+  end
+
+  def load_category
+    @category = Category.find_by id: params[:id]
+    return if @category
+    flash[:danger] = t ".not_found_category"
+    redirect_to root_path
+  end
+
+  def load_parent_category
+    @parent_cats = Category.parent_cats
+  end
+
+  def delete_parent_category parent_category
+    cat_id = parent_category.id
+    if parent_category.destroy
```

Comment:
vì có nhìu xử lý nên chú tìm hiểu về `transaction in rails ` và sử dụng ở đây nhé

Reply:
chỗ ni vì em dùng soft delete nên sẽ không xử lí nhiều việc phải thay đổi category cho các tour nữa. nên em ko dùng transaction chỗ ni nữa. Em sẽ lưu ý để sử dụng ở những chỗ khác ạ.